[quant][pt2] Fix convert in Conv + BN QAT fusion

andrewor14 opened the pull request:
Summary:
Previously, the test for the convert flow in Conv + BN
QAT fusion was not enabled by mistake. However, reenabling this
test uncovered several bugs:

(1) The replaced nodes returned by subgraph rewriter were not
handled correctly. This is because a recent change in the subgraph
rewriter (#100556) fixed only the prepare case but not the convert
case. This commit brings this fix to the convert case as well and
deduplicates some code between the two cases.

(2) When folding BN into conv, we used the wrong arg index to get
the BN eps value. This resulted in an incorrect conv weight.

(3) In FX, we currently do a hack for weighted modules where we
observe the weights once in convert in order to ensure we get the
right shapes for these weight observers. This caused the numerics
to diverge between PT2 and FX. This commit fixes this by skipping
this unnecessary hack for `_convert_to_reference_decomposed_fx`.

(4) Per channel support was simply missing. This commit adds
support for this by matching the quantize_per_channel and
dequantize_per_channel ops in addition to the existing ones.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_numerics

Reviewed By: jerryzh168

Differential Revision: D46097783

jerryzh168 reviewed (approved):
thanks for fixing the errors, sorry about the complicated logic for running v.s. not running `weight_post_process`.. ideally we can make it clearer if we decide to deprecate the fx graph mode quant path (we can remove both non-decomposed and non-reference path)

eellison reviewed (commented):
You could also consider using https://github.com/pytorch/pytorch/blob/main/torch/_inductor/pattern_matcher.py

andrewor14 commented:
> You could also consider using https://github.com/pytorch/pytorch/blob/main/torch/_inductor/pattern_matcher.py

@eellison Thanks for letting me know. Are there any docs where I can read up on how to use this? Also do you know how this compares to the FX subgraph rewriter in terms of functionality? So far we've implemented much of PT2 quantization using the existing subgraph rewriter. If we do switch we would need to be consistent across the code base.

andrewor14 committed:
[quant][pt2] Fix convert in Conv + BN QAT fusion (#102224)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/102224

Previously, the test for the convert flow in Conv + BN
QAT fusion was not enabled by mistake. However, reenabling this
test uncovered several bugs:

(1) The replaced nodes returned by subgraph rewriter were not
handled correctly. This is because a recent change in the subgraph
rewriter (#100556) fixed only the prepare case but not the convert
case. This commit brings this fix to the convert case as well and
deduplicates some code between the two cases.

(2) When folding BN into conv, we used the wrong arg index to get
the BN eps value. This resulted in an incorrect conv weight.

(3) In FX, we currently do a hack for weighted modules where we
observe the weights once in convert in order to ensure we get the
right shapes for these weight observers. This caused the numerics
to diverge between PT2 and FX. This commit fixes this by skipping
this unnecessary hack for `_convert_to_reference_decomposed_fx`.

(4) Per channel support was simply missing. This commit adds
support for this by matching the quantize_per_channel and
dequantize_per_channel ops in addition to the existing ones.

Test Plan: python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_numerics

Reviewed By: jerryzh168

Differential Revision: D46097783

fbshipit-source-id: aa8d30e81fd402a7c1b842eb84b69c16ddd6674b
